yarrp.1
.\"
.\" yarrp.1
.\"
.\" Author: Robert Beverly <rbeverly@cmand.org>
.\"
.\" Copyright (c) 2016-2024 Robert Beverly
.\"                    All rights reserved
.\"
.\"
.Dd Sept 27, 2024
.Dt YARRP 1
.Os
.Sh NAME
.Nm yarrp
.Nd high-speed active IPv4/IPv6 network topology prober
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl hvQT
.Op Fl i Ar target_file
.Op Fl o Ar outfile
.Op Fl r Ar rate
.Op Fl t Ar tr_type
.Op Fl c Ar tr_count
.Op Fl S Ar seed
.Op Fl E Ar instance
.Op Fl p Ar dst_port
.Op Fl b Ar bgp_rib
.Op Fl B Ar blocklist
.Op Fl l Ar min_ttl
.Op Fl m Ar max_ttl
.Op Fl F Ar fill_ttl
.Op Fl n Ar nbr_ttl
.Op Fl s Ar sequential
.Op Fl Z Ar poisson
.Op Fl a Ar src_addr
.Op Fl I Ar interface
.Op Fl M Ar src_mac
.Op Fl G Ar dst_mac
.Op Fl g Ar v6_gran
.Op Fl X Ar v6_eh
.Op Ar subnet(s)
.Sh DESCRIPTION
.Nm
(Yelling at Random Routers Progressively) is a high-speed active
traceroute-style network topology discovery tool.  To achieve its high probing
rates, 
.Nm
is stateless and randomizes the order of probed destinations and
TTLs.  By spreading probes, 
.Nm
distributes load and attempts to avoid 
network rate-limiting.  Yarrp supports both IPv4 and IPv6
and can send probes of any transport type (TCP, UDP-paris, or ICMP-paris).
.Pp
.Sh OPTIONS
The set of IPv4 or IPv6 destination targets to probe may be specified
in one of three ways:
.Bl -tag -width Ds
.It Ar subnet(s)
Probes a target in each /24 (IPv4), or
each /48 (IPv6), of the specified subnets.
.It Fl i Ar target_file
Input list (one address per line) of explicit targets; accepts stdin.
.It Fl Q
Internet-wide scanning.  Probes an address in each /24 (IPv4) or each /48 (IPv6) 
(use with caution).
.El
.Pp
The general options are as follows:
.Bl -tag -width Ds
.It Fl h
print command line options and a synopsis of each.
.It Fl v
verbose (use multiple times to increase verbosity)
.It Fl T
test mode (default: off)
.It Fl o Ar outfile
output file for probing results; accepts stdout. (default: output.yrp)
.It Fl r Ar rate
set packet per second probing rate (default: 10pps)
.It Fl t Ar tr_type
set probe type: TCP_ACK, TCP_SYN, UDP, ICMP, ICMP_REPLY (default: TCP_ACK)
.It Fl c Ar tr_count
set number of traces to issue (default: unlimited)
.It Fl S Ar seed
set permutation random seed (default: timestamp)
.It Fl E Ar instance 
set instance (default: 0)
.It Fl p Ar dst_port
use specified transport destination port (default: 80)
.It Fl a Ar src_addr
set source IP address (default: auto)
.El
.Pp
The target options are as follows:
.Bl -tag -width Ds
.It Fl b Ar bgp_rib
read BGP RIB (Potaroo text format) (default: none)
.It Fl B Ar blocklist
read list of prefixes to skip (default: none)
.El
.Pp
The options to control TTLs probed are:
.Bl -tag -width Ds
.It Fl l Ar min_ttl
set minimum TTL (default: 1)
.It Fl m Ar max_ttl
set maximum TTL, must be a power of 2 (default: 16)
.It Fl F Ar fill_ttl
set fill mode maximum TTL (default: 32)
.It Fl s
send probes sequentially (default: random)
.It Fl n Ar nbr_ttl
enable neighborhood enhancement and set local neighborhood TTL (default: off)
.It Fl Z Ar poisson
choose TTLs from a Poisson distribution with specified lambda (default: uniform)
.El
.Pp
The IPv6-specific options are as follows:
.Bl -tag -width Ds
.It Fl I Ar interface
network interface to use (required)
.It Fl t Ar tr_type
set probe type: ICMP6, UDP6, TCP6_SYN, TCP6_ACK (required)
.It Fl M Ar src_mac
MAC address of source (required if auto discovery fails)
.It Fl G Ar dst_mac
MAC address of gateway router (required if auto discovery fails)
.It Fl g Ar v6_gran
Granularity at which to probe input IPv6 prefixes (default: /50)
.It Fl X Ar v6_eh
Set extension header type to add (default: none)
.El
.Sh OUTPUT
.Nm 
writes probe responses to the specified output file in a delimited
ASCII format as they are received, one response per line.  Because
.Nm
randomizes its probing, results will be similarly randomized.
To determine all of the responses for a single target destination, it
is necessary to filter and collate responses.  The included
yrp2warts utility (provided as both python and C++) performs this reconstitution and produces output
in the standard warts binary format.
.Sh TTLs
By default, 
.Nm
randomly permutes the space of targets and TTLs,
thereby probing each target with TTLs from min_ttl to max_ttl in a random
order.  Note that because of the way
.Nm
permutes the probe order, max_ttl must be a power of two.
.Pp
Four options modify this behavior.  The sequential option
(-s) disables random probing and instead probes sequentially.  The nbr_ttl
option (-n) is an optimization that stops probing low TTLs within the local
neighborhood of the prober once 
.Nm
determines that it is not
discovering any new interfaces within that neighborhood.  In
fill mode (-F), 
.Nm
will probe, up to a maximum TTL of fill_ttl, the next
hop beyond
max_ttl if it receives a response for a probe with TTL
greater than or equal to max_ttl.
.Pp
Finally, the -Z option specifies a lambda parameter for a Poisson
distribution.
.Nm 
will iterate through all TTLs, but the probability of probing
a particular TTL follows a Poisson distribution with the given lambda.
This mode is intended to maximize router discovery yield, as  
the majority of Internet routers are concentrated in a particular
TTL range.
.Sh EXAMPLES
The command:
.Pp
.in +.3i
yarrp -i targets -o test.yrp -r 100
.in -.3i
.Pp
will send TCP_ACK topology probes in a randomly-permuted order to the IPv4 
targets in file "targets" at a rate of 100pps, and write results to
file "test.yrp".
.Pp
The command:
.Pp
.in +.3i
yarrp -o scan.yrp -t ICMP -v -m 16 205.155.0.0/16
.in -.3i
.Pp
will send ICMP topology probes in a randomly-permuted order to all
destinations within the prefix 205.155.0.0/16, from TTL 1 to 16
at the default rate of 10pps.  Verbosity is switched on so that
.Nm
will report probe and response data to stdout.  The results
will be written to the file "scan.yrp".
.Pp
The command:
.Pp
.in +.3i
yarrp -o scan2.yrp -t ICMP -b bgptable.txt 1.0.0.0/8
.in -.3i
.Pp
will send ICMP topology probes in a randomly-permuted order to all
destinations within the prefix 1.0.0.0/8, if the destination
has a route in the BGP routing table "bgptable.txt".  The routing
table file must be plain-text in Potaroo format (the most recent
table is available from https://bgp.potaroo.net/as6447/bgptable.txt).
The results will be written to the file "scan2.yrp".
.Pp
The command:
.Pp
.in +.3i
yarrp -t UDP6 -I eth0 -i targets6 -o test6.yrp
.in -.3i
.Pp
will send UDP probes in a randomly-permuted order to the set of
IPv6 targets in the file "targets6", and write the results to the file 
"test6.yrp".
.Pp
.in -.5i
.Sh SEE ALSO
.Xr yrp2warts.py 1
.Xr warts2yrp.py 1
.Rs
.%A "R. Beverly"
.%T "Yarrp'ing the Internet: Randomized High-Speed Active Topology Discovery"
.%O "Proc. ACM/SIGCOMM Internet Measurement Conference 2016"
.Re
.Rs
.%A "R. Beverly, R. Durairajan, D. Plonka, and J.P. Rohrer"
.%T "In the IP of the Beholder: Strategies for Active IPv6 Topology Discovery"
.%O "Proc. ACM/SIGCOMM Internet Measurement Conference 2018"
.Re
.Rs
.%A "E. C. Rye, and R. Beverly"
.%T "Discovering the IPv6 Network Periphery"
.%O "Proc. Passive and Active Measurement 2020"
.Re
.Rs
.%A "K. Vermeulen, et al."
.%T "Diamond-Miner: Comprehensive Discovery of the Internet's Topology Diamonds"
.%O "Proc. USENIX NSDI 2020"
.Re
.Sh AUTHORS
.Nm
is written by Robert Beverly <rbeverly@cmand.org>.  Ionut Luculescu contributed
support for IPv4 UDP probing.  Eric Gaston contributed support for IPv6 probing.
Oliver Gasser contributed proper rate limiting patches.


CHANGES
yarrp-0.8
---------
* Fix memory leaks
* IPv6 subnet probing
* IPv6 blocklist support
* Better BGP table support
* Misc bug fixes

yarrp-0.7
---------
* Add IPv6 EH support
* Add ICMP MPLS extension support
* FreeBSD verison >=11 fixes

yarrp-0.6
---------
* Add blocklist support
* Option to accept targets from stdin
* Option to export results to stdout 
* Structured yrp output format
* Instance locking
* Fixed rate control
* Add minimum TTL option
* Fast C++ yrp2warts processor
* Bundle libcperm
* Misc stability fixes

yarrp-0.5
---------
* Add fill mode
* Support IPv6 on BSD
* Support multiple simultaneous instances
* Support ICMP echo reply probes

yarrp-0.4
---------
* Multiple IPv6 fixes
* Support IPv6 BGP tables
* Support IPv4 ICMP probing
* Add probe-to-end TTL enhancement
* Microsecond timestamp support
* Record permutation seed in output
* Decoupled probing/receiving support 
* Beta Internet6-wide probing support
* Updated yrp2warts utilities

yarrp-0.3
---------
* IPv6 support
* Require libcperm

yarrp-0.2
---------
* Support UDP probing
* Configurable transport destination port

yarrp-0.1
---------
* Initial public release


yarrp.cpp
/****************************************************************************
 * Copyright (c) 2016-2024 Robert Beverly <rbeverly@cmand.org>
 * All rights reserved.
 *
 * Program:     $Id: yaarp.cpp $
 * Description: yarrp.  https://www.cmand.org/yarrp
 *              indent -i4 -nfbs -sob -nut -ldi1 yarrp.cpp
 *
 * Attribution: R. Beverly, "Yarrp'ing the Internet: Randomized High-Speed
 *              Active Topology Discovery", Proceedings of the ACM SIGCOMM
 *              Internet Measurement Conference, November, 2016
 ***************************************************************************/
#include "yarrp.h"


template < class TYPE >
void
loop(YarrpConfig * config, TYPE * iplist, Traceroute * trace,
     Patricia * tree, Stats * stats) {
    struct in_addr target;
    struct in6_addr target6;
    uint8_t ttl;
    TTLHisto *ttlhisto = NULL;
    Status *status = NULL;
    char ptarg[INET6_ADDRSTRLEN];
    double prob, flip;
    int *asn;

    //adaptive timing to hit target rate
    uint64_t count = 0;
    uint64_t last_count = count;
    double last_time = now();
    uint32_t delay = 0;
    int interval = 0;
    volatile int vi;
    struct timespec ts, rem;
    double send_rate = (double)config->rate;
    const double slow_rate = 50;
    long nsec_per_sec = 1000 * 1000 * 1000;
    long long sleep_time = nsec_per_sec;

    if (config->rate > 0) {
        delay = 10000;
        if (send_rate < slow_rate) {
            //set the inital time difference
            sleep_time = nsec_per_sec / send_rate;
            last_time = now() - (1.0 / send_rate);
        } else {
            //estimate initial rate
            for (vi = delay; vi--;);
            delay *= 1 / (now() - last_time) / (config->rate);
            interval = (config->rate) / 20;
            last_time = now();
        }
    }

    stats->to_probe = iplist->count();
    while (true) {
        /* Grab next target/ttl pair from permutation */
        if (config->ipv6) {
            if ((iplist->next_address(&target6, &ttl)) == 0)
                break;
        } else {
            if ((iplist->next_address(&target, &ttl)) == 0)
                break;
        }
        /* TTL control enforcement */
        ttl += config->minttl;
        if (ttl > config->maxttl) {
            continue;
        }
        /* Running w/ a biased TTL probability distribution */
        if (config->poisson) {
            prob = poisson_pmf(ttl, config->poisson);
            flip = zrand();
            //cout << "TTL: " << (int)ttl << " PMF: " << prob << " flip: " << flip << endl;
            if (flip > prob)
                continue;
        }
        /* Send probe only if outside discovered neighborhood */
        if (ttl < config->ttl_neighborhood) {
            ttlhisto = trace->ttlhisto[ttl];
            if (ttlhisto->shouldProbeProb() == false) {
                //cout << "TTL Skip: " << inet_ntoa(target) << " TTL: " << (int)ttl << endl;
                stats->nbr_skipped++;
                continue;
            }
            ttlhisto->probed(trace->elapsed());
        }
        /* Only send probe if destination is in BGP table */
        if (config->bgpfile or config->blocklist) {
            if (config->ipv6) {
                asn = (int *)tree->get(target6);
            } else {
                asn = (int *)tree->get(target.s_addr);
            }
            if (verbosity >= HIGH)
                inet_ntop(AF_INET, &target, ptarg, INET6_ADDRSTRLEN);
            if (asn == NULL) {
                debug(DEBUG, "BGP Skip: " << ptarg << " TTL: " << (int)ttl);
                stats->bgp_outside++;
                continue;
            }
            if (*asn == 0) {
                debug(HIGH, ">> Address in blocklist: " << ptarg << " TTL: " << (int)ttl);
                continue;
            } else {
                debug(DEBUG, ">> Prefix: " << ptarg << " ASN: " << *asn);
            }
#if 0
                status = (Status *) tree->get(target.s_addr);
                if (status) {
                    status->probed(ttl, trace->elapsed());
                } else {
                    stats->bgp_outside++;
                    continue;
                }
#endif
        }
        /* Passed all checks, continue and send probe */
        if (not config->testing) {
            if (config->ipv6)
                trace->probe(target6, ttl);
            else
                trace->probe(target.s_addr, ttl);
        } else if (verbosity > HIGH) {
            if (config->ipv6)
                trace->probePrint(target6, ttl);
            else
                trace->probePrint(&target, ttl);
        }
        stats->count++;
        /* Progress printer */
        if ((verbosity >= LOW) and
            (iplist->count() > 10000) and
            (stats->count % (iplist->count() / 1000) == 0)) {
            stats->terse();
        }

        /* Calculate sleep time based on scan rate */
        if (config->rate) {
            send_rate = (double)config->rate;
            if (count && delay > 0) {
                if (send_rate < slow_rate) {
                    double t = now();
                    double last_rate = (1.0 / (t - last_time));

                    sleep_time *= ((last_rate / send_rate) + 1) / 2;
                    ts.tv_sec = sleep_time / nsec_per_sec;
                    ts.tv_nsec = sleep_time % nsec_per_sec;
                    while (nanosleep(&ts, &rem) == -1) {
                    }
                    last_time = t;
                } else {
                    for (vi = delay; vi--;);
                    if (!interval || (count % interval == 0)) {
                        double t = now();
                        double multiplier =
                        (double)(count - last_count) /
                        (t - last_time) /
                        (config->rate);
                        uint32_t old_delay = delay;
                        delay *= multiplier;
                        if (delay == old_delay) {
                            if (multiplier > 1.0) {
                                delay *= 2;
                            } else if (multiplier < 1.0) {
                                delay *= 0.5;
                            }
                        }
                        last_count = count;
                        last_time = t;
                    }
                }
            }
        }
        count = stats->count;

        /* Quit if we've exceeded probe count from command line */
        if (stats->count == config->count)
            break;
    }
}

int
sane(YarrpConfig * config) {
    if (not config->testing)
        checkRoot();
    if (config->minttl > config->maxttl)
        fatal("min_ttl must be less than or equal max_ttl");
    if ((config->fillmode > 0) and(config->fillmode < config->maxttl))
        fatal("Fill mode TTL must be larger than max_ttl");
    if (config->ipv6 and not config->testing) {
        if (config->int_name == NULL)
            fatal("IPv6 requires specifying an interface");
    }
    if (config->entire and not config->bgpfile)
        fatal("Entire Internet mode requires BGP table");
    if (config->inlist and config->entire)
        fatal("Cannot run in entire Internet mode with input targets");
    return true;
}

int
main(int argc, char **argv) {
    /* Parse options */
    YarrpConfig config = YarrpConfig();
    config.parse_opts(argc, argv);

    /* Sanity checks */
    sane(&config);

    /* Ensure we're the only Yarrp probing instance on this machine */
    if (config.probe)
        instanceLock(config.instance);

    /* Setup IPv6, if using (must be done before trace object) */
    if (config.ipv6 and not config.testing) {
        if (config.srcmac == NULL || config.dstmac == NULL) {
            LLResolv *ll = new LLResolv();
            ll->gateway();
            ll->mine(config.int_name);
            if (not config.srcmac)
                ll->setSrcMAC(&config.srcmac);
            if (not config.dstmac)
                ll->setDstMAC(&config.dstmac);
            if (config.srcmac == NULL || config.dstmac == NULL) {
                fatal("unable to auto-interpret MAC addresses; use -M, -G");
            }
        }
    }
    /* Init target list (individual IPs, *NOT* subnets) from input file */
    IPList *iplist = NULL;
    if (config.inlist or config.entire) {
        if (config.ipv6)
            iplist = new IPList6(config.maxttl, config.random_scan, config.entire);
        else
            iplist = new IPList4(config.maxttl, config.random_scan, config.entire);
        /* randomize permutation key */
        iplist->setkey(config.seed);
        if (config.inlist)
            iplist->read(config.inlist);
    }
    /* Initialize subnet list and add subnets from args */
    SubnetList *subnetlist = NULL;
    if (not config.entire and not config.inlist and config.probe) {
        if (config.random_scan)
            subnetlist = new RandomSubnetList(config.maxttl, config.granularity);
        else
            subnetlist = new SubnetList(config.maxttl, config.granularity);
        for (int i = optind; i < argc; i++)
            subnetlist->add_subnet(argv[i], config.ipv6);
        if (0 == subnetlist->count())
            config.usage(argv[0]);
    }
    /* Initialize radix trie, if using */
    Patricia *tree = NULL;
    if (config.ipv6) {
        tree = new Patricia(128);
        if (config.blocklist) {
            debug(LOW, ">> Populating IPv6 blocklist: " << config.blocklist);
            tree->populateBlock(AF_INET6, config.blocklist);
        }
        if (config.bgpfile) {
            debug(LOW, ">> Populating IPv6 trie from: " << config.bgpfile);
            tree->populate6(config.bgpfile);
        } else {
            tree->add(AF_INET6, "::/0", 1);
        }
    } else {
        tree = new Patricia(32);
        if (config.blocklist) {
            debug(LOW, ">> Populating IPv4 blocklist: " << config.blocklist);
            tree->populateBlock(AF_INET, config.blocklist);
        }
        if (config.bgpfile) {
                debug(LOW, ">> Populating IPv4 trie from: " << config.bgpfile);
            //tree->populateStatus(config.bgpfile);
            tree->populate(config.bgpfile);
        } else {
            tree->add("0.0.0.0/0", 1);
        }
    }
    /* Initialize traceroute engine, if not in test mode */
    Stats *stats = new Stats();
    Traceroute *trace = NULL;
    if (config.ipv6)
        trace = new Traceroute6(&config, stats);
    else
        trace = new Traceroute4(&config, stats);

    trace->addTree(tree);

    /* Open output */
    if (config.receive) {
        config.dump();
        /* unlock so listener thread starts */
        trace->unlock();
    }
    /* Start listener if we're only in receive mode */
    if ((not config.probe) and config.receive) {
        if (config.ipv6)
            listener6(trace);
        else
            listener(trace);
    }
    if (config.probe) {
        debug(LOW, ">> Probing begins.");
        if (config.entire or config.inlist) {
            /* individual IPs from input file or entire mode */
            loop(&config, iplist, trace, tree, stats);
        } else {
            /* using subnets from args */
            loop(&config, subnetlist, trace, tree, stats);
        }
    }
    if (config.receive) {
        debug(LOW, ">> Waiting " << SHUTDOWN_WAIT << "s for outstanding replies...");
        sleep(SHUTDOWN_WAIT);
    }
    /* Finished, cleanup */
    if (config.receive) {
        if (config.output and not config.testing)
            stats->dump(trace->config->out);
        else
            stats->dump(stdout);
    }
    delete stats;
    delete trace;
    if (tree)
        delete tree;
    if (iplist)
        delete iplist;
    if (subnetlist)
        delete subnetlist;
}
